modules/java/native/java-grammar.ym

```diff
@@ -86,6 +87,14 @@ java_dest_custom_options
   |
   ;
 
+java_dest_empty_options
+  : java_dest_empty_option java_dest_empty_options
+  |
+  ;
+
+java_dest_empty_option
+  : LL_STRING { ;  }
+
```

Comment:
Hi,

I could live with this limitation, but I don't know @juhaszviktor 's
opinion.

I would btw try to get rid off the option() syntax and use stray name-value
pairs in the block itself, if possible.

e.g. java(foo(bar) baz(bax)) and that way we could generate keyword-args
style block options into the resulting configuration snipped, e.g.

block source stuff(foo(bar) baz(bax) **kwargs) {
    java(arg(`foo`) arg2(`baz`) **kwargs)
}

I used the Python kwargs markup here with the double asterisk (not
necessarily the best idea in this case), but you get the idea. If this was
implemented, optional arguments would simply be optional arguments.


-- 
Bazsi

On Mon, Jul 13, 2015 at 2:59 PM, aneutrals <notifications@github.com> wrote:

> In modules/java/native/java-grammar.ym
> <https://github.com/balabit/syslog-ng/pull/556#discussion_r34459813>:
>
> > @@ -86,6 +87,14 @@ java_dest_custom_options
> >    |
> >    ;
> >
> > +java_dest_empty_options
> > +  : java_dest_empty_option java_dest_empty_options
> > +  |
> > +  ;
> > +
> > +java_dest_empty_option
> > +  : LL_STRING { ;  }
> > +
>
> I included an example in my original post, but I will try to make it
> clearer here.
>
> If I have a java destination that optionally takes a parameter "foo", that
> is, behaves differently when the dictionary of parameters passed to Java
> includes "foo", with an associated value, I cannot write an SCL block for
> that destination currently without forcing "foo" to have a value always -
> that is, turning it from an optional option to a required option.
>
> If in my scl block, I do not define a default value for "foo", parsing the
> block on startup will fail, because the grammar expects an option to have a
> value.
>
> block destination test_block(foo()){
>   java(
>     class_name("sampleclass")
>     class_path("/home/user/sampleclass/")
>     option("foo",`foo`)
>   );
> };
> #Syslog-ng would fail to parse this because a value wasn't given to foo
>
> To appease the parser, I *could* include a default value for foo in the
> block definition, but that would mean "foo" will always be in the
> dictionary of options passed to Java - it no longer is an "optional" value.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/556/files#r34459813>.
>